src/tensorwaves/optimizer/callbacks.py

```diff
@@ -0,0 +1,92 @@
+"""Collection of loggers that can be inserted into an optimizer as callback."""
+
+from abc import ABC, abstractmethod
+from datetime import datetime
+from typing import Any, Iterable, List, Optional
+
+import tensorflow as tf
+from tqdm import tqdm
+
+
+class Callback(ABC):
+    @abstractmethod
+    def __init__(self, **kwargs: Any) -> None:
+        pass
+
+    @abstractmethod
+    def __call__(self, **kwargs: Any) -> None:
```

Comment:
I don't like kwargs in an interface, as it blurs the whole interface. Whats the reasoning for that? I mean there is only as rather specific set of information that can be passed from the optimizer to the callback (currently we only can think of parameters and the estimator value)

Reply:
I also don't really like `kwargs` and initially had `parameters` and `estimator_value` in the interface. But then I had to use those in the `ProgressBar` as well without them being used, so then I wondered whether it's a good idea to enforce these arguments. Wouldn't such an interface be too tailored to how it is  used in the current `Optimizer` implementation (currently only `Minuit2`)?

As for using `kwargs` at all: I would say you would need that if you want to register additional information. So for me the question is more whether or not to enforce `parameters` and `estimator_value`.

For now I see a `Callback` as a more abstract procedure: a hook that can be called and that can be finalised.

Reply:
> I mean there is only as rather specific set of information that can be passed from the optimizer to the callback (currently we only can think of parameters and the estimator value)

An argument for this could be that the `callbacks` module is located under `optimizer`

Reply:
> I also don't really like `kwargs` and initially had `parameters` and `estimator_value` in the interface. But then I had to use those in the `ProgressBar` as well without them being used, so then I wondered whether it's a good idea to enforce these arguments. Wouldn't such an interface be too tailored to how it is used in the current `Optimizer` implementation (currently only `Minuit2`)?
> 
That is true. But that's the whole point of an interface. That is standardizes the procedure and allow to "hot swap" modules and things will continue to work. In other words this kwargs "interface" violates the Liskov Substitution Principle (btw did you get any error message on that?). As an example, lets say one Callback requires some argument `estimator_value`, another Callback maybe `log_likelihood` (or whatever). But how do you know in the Optimizer, which Callbacks need which arguments? You can't unless you define an interface with a concrete specification.

> As for using `kwargs` at all: I would say you would need that if you want to register additional information. So for me the question is more whether or not to enforce `parameters` and `estimator_value`.
> 
> For now I see a `Callback` as a more abstract procedure: a hook that can be called and that can be finalised.

If we need to be more flexible for more custom values. Then maybe the access to that info has to be given "by reference" beforehand. So that the callback receives the estimator, parameters and maybe some part of the optimizer and can extract all of the relevant info from there. Btw some info that might be interesting to be logged in the future would be the gradient info (if available)

Reply:
Good points :+1: Here are two versions that may fix it:
15b1dbe (enforce `parameters` and `estimator_value`)
b69e549 (remove kwargs as well).

Reply:
> Then maybe the access to that info has to be given "by reference" beforehand.

How about defining a `FitResult` class? (Would be a follow-up issue)

Reply:
> Good points  Here are two versions that may fix it:
> [15b1dbe](https://github.com/ComPWA/tensorwaves/commit/15b1dbed73e81c02b66e332a11688c93cbf28abc) (enforce `parameters` and `estimator_value`)
> [b69e549](https://github.com/ComPWA/tensorwaves/commit/b69e549d7c422c6bb2418a8e005ef414746d55c2) (remove kwargs as well).

Looks good. We will have to see if we can keep this interface stable. If it requires to much customized behaviors of Optimizers to loggers, then we have to reconsider this design...

Reply:
> > Then maybe the access to that info has to be given "by reference" beforehand.
> 
> How about defining a `FitResult` class? (Would be a follow-up issue)

In ComPWA (c++ codebase) this FitResult was templated. Because this output is strongly dependent on/customized by the Optimizer that is used.